File: client-android/lib/src/main/kotlin/stasis/client_android/lib/encryption/secrets/UserAuthenticationPassword.kt
package stasis.client_android.lib.encryption.secrets

import okio.ByteString
import stasis.client_android.lib.model.server.users.UserId
import java.security.MessageDigest
import java.util.Base64
import java.util.concurrent.atomic.AtomicBoolean

sealed class UserAuthenticationPassword : Secret() {
    abstract val user: UserId
    abstract fun extract(): String
    abstract fun digested(): String

    data class Hashed(
        override val user: UserId,
        private val hashedPassword: ByteString
    ) : UserAuthenticationPassword() {
        private val extracted: AtomicBoolean = AtomicBoolean(false)

        override fun extract(): String {
            val alreadyExtracted = extracted.getAndSet(true)

            return if (alreadyExtracted) {
                throw IllegalStateException("Password already extracted")
            } else {
                Base64.getUrlEncoder().withoutPadding().encodeToString(hashedPassword.toByteArray())
            }
        }

        override fun digested(): String =
            Base64.getUrlEncoder().withoutPadding().encodeToString(
                MessageDigest.getInstance("SHA-512").digest(hashedPassword.toByteArray())
            )
    }

    data class Unhashed(
        override val user: UserId,
        private val rawPassword: ByteString
    ) : UserAuthenticationPassword() {
        private val extracted: AtomicBoolean = AtomicBoolean(false)

        override fun extract(): String {
            val alreadyExtracted = extracted.getAndSet(true)

            return if (alreadyExtracted) {
                throw IllegalStateException("Password already extracted")
            } else {
                rawPassword.utf8()
            }
        }

        override fun digested(): String =
            Base64.getUrlEncoder().withoutPadding().encodeToString(
                MessageDigest.getInstance("SHA-512").digest(rawPassword.toByteArray())
            )
    }
}

File: client/src/main/scala/stasis/client/encryption/secrets/UserAuthenticationPassword.scala
package stasis.client.encryption.secrets

import java.util.concurrent.atomic.AtomicBoolean

import org.apache.pekko.util.ByteString

import stasis.shared.model.users.User
import stasis.shared.secrets.DerivedPasswords

// doc - sent to auth provider
sealed trait UserAuthenticationPassword extends Secret {
  def extract(): String
  def digested(): String
}

object UserAuthenticationPassword {
  final case class Hashed(
    user: User.Id,
    private val hashedPassword: ByteString
  ) extends UserAuthenticationPassword {
    private val extracted: AtomicBoolean = new AtomicBoolean(false)

    @SuppressWarnings(Array("org.wartremover.warts.Throw"))
    override def extract(): String = {
      val alreadyExtracted = extracted.getAndSet(true)

      if (alreadyExtracted) {
        throw new IllegalStateException("Password already extracted")
      } else {
        DerivedPasswords.encode(hashedPassword)
      }
    }

    override def digested(): String = DerivedPasswords.digest(password = hashedPassword)
  }

  object Hashed {
    def apply(user: User.Id, hashedPassword: ByteString): Hashed =
      new Hashed(user, hashedPassword)
  }

  final case class Unhashed(
    user: User.Id,
    private val rawPassword: ByteString
  ) extends UserAuthenticationPassword {
    private val extracted: AtomicBoolean = new AtomicBoolean(false)

    @SuppressWarnings(Array("org.wartremover.warts.Throw"))
    override def extract(): String = {
      val alreadyExtracted = extracted.getAndSet(true)

      if (alreadyExtracted) {
        throw new IllegalStateException("Password already extracted")
      } else {
        rawPassword.utf8String
      }
    }

    override def digested(): String = DerivedPasswords.digest(password = rawPassword)
  }

  object Unhashed {
    def apply(user: User.Id, rawPassword: ByteString): Unhashed =
      new Unhashed(user, rawPassword)
  }
}
